### PR TITLE
Gaussian unitary

### PR DESCRIPTION
**Pull request by nquesada**
**Context:**
Any arbitrary sequence of Gaussian operations can always be compiled into an interferometer, followed by single mode squeezing, followed by another interferometer, and finally followed by displacements. 

**Description of the Change:**
This PR implements a Gaussian compiler that takes an arbitrary sequence of Gaussian operations and returns a `GaussianUnitary` + a sequence of `Dgates`.

**Benefits:**
Compilation of arbitrarily long Gaussian circuits into their most compact form.

**Possible Drawbacks:**
It requires version 0.9.0 of The Walrus

**Related GitHub Issues:**

### CODE REVIEW

**Comment by nquesada** [comment]
Why are `GraphEmbed` and `BipartiteGraphEmbed` considered decompositions and not state preparations? @josh146, @co9olguy ?

**Comment by josh146** [comment]
> Why are `GraphEmbed` and `BipartiteGraphEmbed` considered decompositions and not state preparations?

I suppose they should be, like how `GaussianState` is both a decomposition and a state preparation.

By the way, you'll need to add tests for the new compilation method - to start it should be fine to copy and existing test (maybe test_circuitspecs_chip0.py), and modify/add tests where needed

File: tests/frontend/test_circuitspecs_gaussianunitary.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    with circuit.context as q:
+        ops.GraphEmbed(A) | q
+        ops.BipartiteGraphEmbed(B) | q
+        ops.Pgate(0.1) | q[1]
+        ops.CXgate(0.2) | (q[0], q[1])
+        ops.MZgate(0.4, 0.5) | (q[2], q[3])
+        ops.Fourier | q[0]
+        ops.Xgate(0.4) | q[1]
+        ops.Zgate(0.5) | q[3]
+    compiled_circuit = circuit.compile("gaussian_unitary")
```
Comment (josh146) [inline tests/frontend/test_circuitspecs_gaussianunitary.py]: this is awesome

File: strawberryfields/utils.py [hunk 2]
```diff
@@ -387,6 +388,9 @@ def random_symplectic(N, passive=False, block_diag=False):
         block_diag (bool): If True, uses passive Gaussian transformations that are orthogonal
             instead of unitary. This implies that the positions :math:`q` do not mix with
             the momenta :math:`p` and thus the symplectic operator is block diagonal
+        scale (float): Sets the scale of the random values used as squeezing parameters.
+            They will range from 0 to :math:`\sqrt{2}`*scale
```
Comment (josh146) [inline strawberryfields/utils.py]:
```suggestion
            They will range from 0 to :math:`\sqrt{2}\texttt{scale}`
```

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        Raises:
+            CircuitError: the circuit does not correspond to a Gaussian unitary
+        """
+
+        # Check which modes are actually being used
+        used_modes = []
+        for operations in seq:
+            modes = [modes_label.ind for modes_label in operations.reg]
+            used_modes.append(modes)
+        # pylint: disable=consider-using-set-comprehension
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]: How come you don't use set comprehension here?

**Comment by josh146** [comment]
Also don't forget to update the changelog!

File: strawberryfields/circuitspecs/gaussian_unitary.py (hunk 3 above)
Comment (nquesada) [reply strawberryfields/circuitspecs/gaussian_unitary.py]: I don't think it makes a difference.

File: tests/frontend/test_circuitspecs_gaussianunitary.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    mean1 = eng1.run(compiled_circuit).state.means()
+    assert np.allclose(cv, cv1)
+    assert np.allclose(mean, mean1)
+
+
+
+@pytest.mark.parametrize("depth", [1, 3, 6])
+@pytest.mark.parametrize("width", [5, 10, 15])
+def test_displacements_only(depth, width):
+    """Tests that a circuit and its compiled version produce the same Gaussian state"""
```
[the remark this answers is not in the document]
Comment (nquesada) [reply tests/frontend/test_circuitspecs_gaussianunitary.py]: I have updated the docstring

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+import numpy as np
+from strawberryfields.program_utils import Command
+from strawberryfields import ops
+from strawberryfields.parameters import par_evaluate
+from thewalrus.symplectic import expand_vector, expand, rotation, squeezing, two_mode_squeezing, interferometer, beam_splitter
+from .circuit_specs import CircuitSpecs
+
+class GaussianUnitary(CircuitSpecs):
+    """Circuit specifications for the Gaussian Unitary compiler
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]:
```suggestion
    """Compiler to arrange a Gaussian quantum circuit into the canonical Symplectic form.
    
    This compile specification checks whether the circuit can be implemented as a sequence of
    Gaussian operations. If so, it arranges them in the canonical order with displacement at the end.
    After compilation, the circuit will consist of at most two operations, a :class:`~.GaussianTransform`
    and a :class:`~.Dgate`.
    
    This compiler can be accessed by calling :meth:`.Program.compile` with `'gaussian_unitary'` specified.
```

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+        **Example:**
+
+        The following strawberryfield command sequence:
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]:
```suggestion
        Consider the following Strawberry Fields program, compiled using the `'gaussian_unitary'` compiler:

```

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            import strawberryfields as sf
+
+            circuit = sf.Program(1)
+            with circuit.context as q:
+                Xgate(0.4) | q[0]
+                Zgate(0.5) | q[0]
+                Sgate(0.6) | q[0]
+                Dgate(1.0+2.0j) | q[0]
+                Rgate(0.3) | q[0]
+                Sgate(0.6, 1.0) | q[0]
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]:
```suggestion
                Sgate(0.6, 1.0) | q[0]

```

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            compiled_circuit = circuit.compile("gaussian_unitary")
+
+        We can now print the compiled circuit:
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]:
```suggestion
        We can now print the compiled circuit, consisting of one
        :class:`~.GaussianTransform` and one :class:`~.Dgate`:
```

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+        >>> compiled_circuit.print()
+        GaussianTransform([[ 0.3543 -1.3857]
+                           [-0.0328  2.9508]]) | (q[0])
+        Dgate(-1.151+3.91j, 0) | (q[0])
+
+        will produce a compiled circuit consisting of one GaussianTransform and one Dgate
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]:
```suggestion
```

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
```
Comment (josh146) [inline strawberryfields/circuitspecs/gaussian_unitary.py]:
```suggestion
```

**Pull request merged**